x/fungible/keeper/msg_server_update_system_contract.go

```diff
@@ -0,0 +1,74 @@
+package keeper
+
+import (
+	"context"
+	"math/big"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	ethcommon "github.com/ethereum/go-ethereum/common"
+	"github.com/zeta-chain/protocol-contracts/pkg/contracts/zevm/systemcontract.sol"
+	"github.com/zeta-chain/protocol-contracts/pkg/contracts/zevm/zrc20.sol"
+	"github.com/zeta-chain/zetacore/common"
+	"github.com/zeta-chain/zetacore/x/fungible/types"
+	zetaObserverTypes "github.com/zeta-chain/zetacore/x/observer/types"
+)
+
+func (k Keeper) UpdateSystemContract(goCtx context.Context, msg *types.MsgUpdateSystemContract) (*types.MsgUpdateSystemContractResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	if msg.Creator != k.zetaobserverKeeper.GetParams(ctx).GetAdminPolicyAccount(zetaObserverTypes.Policy_Type_deploy_fungible_coin) {
+		return nil, sdkerrors.Wrap(sdkerrors.ErrUnauthorized, "Deploy can only be executed by the correct policy account")
+	}
+	newSystemContractAddr := ethcommon.HexToAddress(msg.NewSystemContractAddress)
+	if newSystemContractAddr == (ethcommon.Address{}) {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "invalid system contract address (%s)", msg.NewSystemContractAddress)
+	}
+
+	// update contracts'
+	zrc20ABI, err := zrc20.ZRC20MetaData.GetAbi()
+	if err != nil {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "failed to get zrc20 abi")
+	}
+	sysABI, err := systemcontract.SystemContractMetaData.GetAbi()
+	if err != nil {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "failed to get system contract abi")
+	}
+	foreignCoins := k.GetAllForeignCoins(ctx)
+	tmpCtx, commit := ctx.CacheContext()
+	for _, fcoin := range foreignCoins {
+		zrc20Addr := ethcommon.HexToAddress(fcoin.Zrc20ContractAddress)
+		if zrc20Addr == (ethcommon.Address{}) {
+			k.Logger(ctx).Error("invalid zrc20 contract address", "address", fcoin.Zrc20ContractAddress)
+			continue
+		}
+		_, err = k.CallEVM(tmpCtx, *zrc20ABI, types.ModuleAddressEVM, zrc20Addr, BigIntZero, nil, true, "updateSystemContractAddress", newSystemContractAddr)
+		if err != nil {
+			return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "failed to call zrc20 contract method updateSystemContractAddress (%s)", err.Error())
+		}
+		if fcoin.CoinType == common.CoinType_Gas {
+			_, err = k.CallEVM(tmpCtx, *sysABI, types.ModuleAddressEVM, newSystemContractAddr, BigIntZero, nil, true, "setGasCoinZRC20", big.NewInt(fcoin.ForeignChainId), zrc20Addr)
+			if err != nil {
+				return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "failed to call system contract method setGasCoinZRC20 (%s)", err.Error())
+			}
+			_, err = k.CallEVM(tmpCtx, *sysABI, types.ModuleAddressEVM, newSystemContractAddr, BigIntZero, nil, true, "setGasZetaPool", big.NewInt(fcoin.ForeignChainId), zrc20Addr)
+			if err != nil {
+				return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "failed to call system contract method setGasZetaPool (%s)", err.Error())
```

Comment:
Same as above, I don't think these errors should be `ErrInvalidRequest`.
I personally think we should define a custom SDK error for the different situation for debugging purposes